src/coreclr/nativeaot/Runtime/DebugHeader.cpp

```diff
@@ -0,0 +1,294 @@
+    // It can also be changed if we want to make a breaking change so drastic that
+    // earlier debuggers should treat the module as if it had no .Net runtime at all.
+    // If the cookie is valid a debugger is safe to assume the Major/Minor version fields
+    // will follow, but any contents beyond that depends on the version values.
+    // The cookie value is currently set to 0x4E 0x41 0x44 0x48 (NADH in ascii)
+    const uint8_t Cookie[4] = { 0x4E, 0x41, 0x44, 0x48 };
+
+    // This counter can be incremented to indicate breaking changes
+    // This field must be encoded little endian, regardless of the typical endianess of
+    // the machine
+    const uint16_t MajorVersion = 1;
+
+    // This counter can be incremented to indicate back-compatible changes
+    // This field must be encoded little endian, regardless of the typical endianess of
+    // the machine
+    const uint16_t MinorVersion = 0;
+
+    // These flags must be encoded little endian, regardless of the typical endianess of
+    // the machine. Ie Bit 0 is the least significant bit of the first byte.
+    // Bit 0 - Set if the pointer size is 8 bytes, otherwise pointer size is 4 bytes
+    // Bit 1 - Set if the machine is big endian
+    // The high 30 bits are reserved. Changes to these bits will be considered a
+    // back-compatible change.
+    const uint32_t Flags = sizeof(void*) == 8 ? 0x1 : 0x0;
+
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding1 = 0;
+
+    // Header pointers below here are encoded using the defined pointer size and endianess
+    // specified in the Flags field. The data within the contracts they point to also uses
+    // the same pointer size and endianess encoding unless otherwise specified.
+
+    // The length of the DebugTypeEntries array
+    #define DEBUG_TYPES_ARRAY_SIZE 101
+    const uint32_t DebugTypeEntriesArraySize = DEBUG_TYPES_ARRAY_SIZE;
+
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding2 = 0;
+
+    // An array describing important types and their offsets
+    DebugTypeEntry DebugTypeEntries[DEBUG_TYPES_ARRAY_SIZE];
+
+    // The length of the GlobalEntries array
+    #define GLOBALS_ARRAY_SIZE 5
+    const uint32_t GlobalEntriesArraySize = GLOBALS_ARRAY_SIZE;
+    
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding3 = 0;
+
+    // An array that contains pointers to important globals
+    GlobalValueEntry GlobalEntries[GLOBALS_ARRAY_SIZE];
+};
+
+extern "C" NativeAOTRuntimeDebugHeader g_NativeAOTRuntimeDebugHeader = {};
```

Comment:
Would it make sense to have non-NativeAOT specific name for this? E.g. `DotNetRuntimeDebugHeader`.

I would love to see this unified accross all runtime variants eventually.

Reply:
I also would like to see this approach unified across all the runtimes, I was originally thinking that they would have their own contracts but there's no reason not to have the same name for the type of the contract itself.

I would like to keep the export name as "g_NativeAOTRuntimeDebugHeader" though so we can differentiate runtimes by just checking the export name. So then coreclr can have g_CoreCLRRuntimeDebugHeader, etc.

Reply:
We like to talk about NativeAOT project as a flavor of CoreCLR runtime. The place we want to get to is there is a continuum of runtime configurations to choose from, not distinct runtimes.

Reply:
Is the concern over API surface area? For a tool like SOS which is inspecting runtime level data structures we will have to be aware which runtime flavor we are running against for things like whether to look up the EEType in the symbols file, or to use CLR metadata to get type information.

I'm happy to change it to g_DotNetRuntimeDebugHeader, but then I would have to either differentiate on the Cookie or some define we put in the contract. That doesn't feel meaningfully different to me other than the API surface.

Reply:
The places that use this should differentiate on contract type, not based on runtime type.

For example, stresslog or GC contract is same between all runtimes, no need to differentiate today.

For type system contract, there are multiple choices - e.g. try JIT-based type system contract first, and if it is not available try no-JIT AOT-based contract second.